rust/noosphere-ns/src/records.rs

```diff
@@ -0,0 +1,318 @@
+///
+/// # Serialization
+///
+/// When transmitting records across the network, they're encoded
+/// as JSON-formatted UTF-8 strings. The [NSRecord::to_bytes]
+/// and [NSRecord::from_bytes] methods respectively handle the
+/// serialization and deserialization in this format.
+///
+/// Fields are mapped by:
+///
+/// `cid` => `"cid": String`
+/// `expires` => `"exp": Number`
+///
+/// An example of the serialized payload structure and
+/// conversion looks like:
+///  
+/// ```
+/// use noosphere_ns::NSRecord;
+/// use cid::{Cid, multihash::{Code, MultihashDigest}};
+/// use std::str::FromStr;
+///
+/// let cid_str = "bafkreibme22gw2h7y2h7tg2fhqotaqjucnbc24deqo72b6mkl2egezxhvy";
+/// let expires = 1667262626;
+///
+/// let record = NSRecord::new(Cid::from_str(cid_str).unwrap(), expires);
+/// assert_eq!(record.cid.to_string(), cid_str);
+/// assert_eq!(record.expires, expires);
+///
+/// let bytes = record.to_bytes().unwrap();
+/// assert_eq!(&String::from_utf8(bytes.clone()).unwrap(), "{\"cid\":\"bafkreibme22gw2h7y2h7tg2fhqotaqjucnbc24deqo72b6mkl2egezxhvy\",\"exp\":1667262626}");
+/// assert_eq!(NSRecord::from_bytes(bytes).unwrap(), record);
+/// ```
+#[derive(Debug, Eq, PartialEq, Clone)]
+pub struct NSRecord {
+    pub cid: Cid,
+    pub expires: u64,
+}
+
+impl NSRecord {
+    /// Creates a new [NSRecord].
+    pub fn new(cid: Cid, expires: u64) -> Self {
+        Self { cid, expires }
+    }
+
+    /// Creates a new [NSRecord] with an expiration `ttl` seconds from now.
+    pub fn new_from_ttl(cid: Cid, ttl: u64) -> Result<Self, Error> {
+        let expires: u64 = SystemTime::now()
+            .checked_add(Duration::new(ttl, 0))
+            .ok_or_else(|| anyhow!("Duration overflow."))?
+            .duration_since(UNIX_EPOCH)
+            .map_err(|e| anyhow!(e.to_string()))?
+            .as_secs();
+        Ok(Self { cid, expires })
+    }
+
+    /// Creates a new [NSRecord] from serialized bytes. See [NSRecord]
+    /// for serialization details.
+    pub fn from_bytes(bytes: Vec<u8>) -> Result<Self, Error> {
+        let string = String::from_utf8(bytes).map_err(|e| anyhow!(e.to_string()))?;
+        serde_json::from_str(&string).map_err(|e| anyhow!(e.to_string()))
+    }
+
+    /// Serializes the record into bytes. See [NSRecord] for
+    /// serialization details.
+    pub fn to_bytes(&self) -> Result<Vec<u8>, Error> {
+        let bytes = serde_json::to_vec(self).map_err(|e| anyhow!(e.to_string()))?;
+        Ok(bytes)
+    }
+
+    /// Validates the [NSRecord] based off of its expiration time
+    /// compared to the current system time.
+    pub fn is_expired(&self) -> bool {
+        match SystemTime::now().duration_since(UNIX_EPOCH) {
+            Ok(duration) => duration.as_secs() >= self.expires,
+            Err(_) => false,
+        }
+    }
+}
+
+impl fmt::Display for NSRecord {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> Result<(), fmt::Error> {
+        match serde_json::to_string(self) {
```

Comment:
Any issues with implementing this display as JSON? No expectations on how display strings are rendered, but the JSON form fits well to reduce the amount of potential formats this data can be displayed.